slick/src/main/scala/slick/util/ReadAheadIterator.scala

```diff
@@ -49,3 +43,11 @@ trait ReadAheadIterator[+T] extends BufferedIterator[T] {
     } else throw new NoSuchElementException("next on empty iterator");
   }
 }
+
+object ReadAheadIterator {
+
+  /** Feature implemented in Scala library 2.12 this maintains functionality for 2.11 */
+  implicit class headOptionReverseCompatibility[T](readAheadIterator: ReadAheadIterator[T]){
+    def headOption : Option[T] = if (readAheadIterator.hasNext) Some(readAheadIterator.head) else None
```

Comment:
`headOption` should not throw an Exception on an empty Iterator. That's the whole point of having `headOption` in addition to `head`. I don't see how the old implementation would have done that but if it did, I'd consider it a bug that should be fixed.